### DISCUSSION TITLE
Typing of FilterQuery<T> and EntityManager.find

### DISCUSSION THREAD

**Discussion by juan-restrepo** [mikro-orm/mikro-orm#1993] [Q&A]
I’m having issues adding `where` clauses and typing:

I have a model called `UserTask`. I’m implementing a static method on it that will allow me wrap the logic of building the query filter for it. If I do:

```TS
    // import { EntityManager, EntityRepository } from '@mikro-orm/postgresql'; (I’m on 4.5.6 version of everything)
    // em: EntityManager

    const filterQuery: FilterQuery<UserTask> = { // If I don't type it here, it won't let me add optional fields
      cognitoUserSub, // This is the only "required" field I need in the query filter
    };
    if (!_.isUndefined(taskTypeId)) {
      const userTaskTypeRepository: EntityRepository<UserTaskType> = em.getRepository(UserTaskType);
      const userTaskTypeReference: UserTaskType = userTaskTypeRepository.getReference(
        typeof taskTypeId === 'number' ? taskTypeId : parseInt(taskTypeId, 10),
      );
      filterQuery.userTaskType = userTaskTypeReference;
    }
    if (!includeExpiredTasks) {
      filterQuery.expirationDate = { $gt: new Date() };
    }
    if (!includeCompletedTasks) {
      filterQuery.isCompleted = false;
    }
    const tasks: UserTask[] = await em.find(UserTask, filterQuery);
```

then the last line is very unhappy because it says that the returned value from the `.find` method in the entity manager is missing some things that `UserTask[]` expects:

```
is missing the following properties from type 'UserTask': isInitialized, populated, toReference, toObject, and 4 more.
```

If I just do something like:

```TS
const tasks: UserTask[] = await em.find(UserTask, { cognitoUserSub });
```

Everything works fine. Somehow, the typing of `filterQuery` as a `FilterQuery<UserTask>` is breaking the typings

My current solution is to do:
```TS
    const filterQuery: any = {
      cognitoUserSub,
    };
```
at the top, but this feels like it’s just a hack and defying the purpose of actually having typings.

**Comment by B4nan** [comment]
Known issue, but very hard to get around while maintaining how the `FilterQuery` currently works (it allows to pass raw PKs too, which is the root cause here, as well as arrays of PKs). We should probably add a new type that will only work with objects but will allow composition, maybe even use the name `FilterQuery` and internally use something else. Will try to spend some time on it before v5 gets out.
